feat(Foxford): add presence

cramatsu opened the pull request:
## Description 
Adding `presence` for **[https://foxford.ru](https://foxford.ru)**

**Note to the founders of this project:**
This online school is the most visited in the Russian Federation (more than `300k` students).
Peace ✌🏿

## Acknowledgements
- [x] I read the [Presence Guidelines](https://github.com/PreMiD/Presences/blob/main/.github/CONTRIBUTING.md)
- [x] I linted the code by running `yarn format`
- [x] The PR title follows the repo's [commit conventions](https://github.com/PreMiD/Presences/blob/main/.github/COMMIT_CONVENTION.md)

## Screenshots
<details>
<summary> Proof showing the creation/modification is working as expected </summary>

![](https://i.imgur.com/p4UhMG4.png)
![](https://i.imgur.com/AXJDhFg.png)

<!-- 
    Screenshots of the presence settings (if applicable) and at least TWO screenshots of the presence displaying correctly
    Including these screenshots will assist the reviewing processes thus speeding up the process of the pull request being merged
-->



</details>

plskz commented on websites/C/Craiyon/dist/metadata.json:
```diff
@@ -14,5 +14,10 @@
 	"category": "other",
-	"tags": ["ai", "dalle", "image", "generation"]
-}
+	"tags": [
+		"ai",
+		"dalle",
+		"image",
+		"generation"
+	]
+}
```
revert

plskz commented on websites/H/Heroku/dist/metadata.json:
```diff
@@ -24,7 +24,11 @@
 	"color": "#211746",
 	"category": "other",
-	"tags": ["hosting", "application", "web"],
+	"tags": [
+		"hosting",
+		"application",
+		"web"
+	],
```
revert

plskz commented on websites/R/Raven Code/dist/metadata.json:
```diff
@@ -20,4 +20,4 @@
 		"raven-code"
 	]
-}
+}
```
revert

plskz commented on websites/S/Skype/dist/metadata.json:
```diff
@@ -42,4 +42,4 @@
 		}
 	]
-}
+}
```
revert

plskz commented on websites/W/War Thunder/dist/metadata.json:
```diff
@@ -28,4 +28,4 @@
 		"battleship"
 	]
-}
+}
```
revert

cramatsu replied:
Understood

plskz reviewed (changes requested):
only commit your `websites/F/Foxfort` folder

cramatsu commented:
> only commit your `websites/F/Foxfort` folder

Done

plskz commented on websites/F/Foxford/presence.ts:
```diff
@@ -0,0 +1,73 @@
+			break;
+	}
+
+	if (presenceData.details) presence.setActivity(presenceData);
+	else presence.setActivity();
```
just a suggestion, same result

```suggestion
	presence.setActivity(presenceData);
```

DarkVillager replied:
Nah keept it as it is

cramatsu replied:
> Nah keept it as it is

Okay, I won't touch these strings

plskz reviewed (commented):
can you support more pages?

cramatsu commented:
> can you support more pages? 

I can, I'll try to do it today.

cramatsu commented:
Guys, look at my commits. About `foxford.ru`: I think the current paths are enough, over time I will update `foxford.ru`

cramatsu commented:
I thought maybe we should split it into two prs?

theusaf commented:
> I think maybe we should split this into two prs?

Yes, the two changes (adding Foxford and modifying jut.su) should be in two different PRs.

cramatsu commented:
> > I think maybe we should split this into two prs?
> 
> Yes, the two changes (adding Foxford and modifying jut.su) should be in two different PRs.

So, I have undone all non-Foxford related changes.
Now, is PR ready to merge?

plskz commented on websites/F/Foxford/dist/metadata.json:
```diff
@@ -0,0 +1,19 @@
+	"url": "foxford.ru",
+	"version": "1.0.0",
+	"logo": "https://i.imgur.com/41WDw6H.jpeg",
+	"thumbnail": "https://i.imgur.com/41WDw6H.jpeg",
```
The thumbnail **should** preferably be a [wide promotional card](https://i.imgur.com/3QfIc5v.jpg) or a [screenshot](https://i.imgur.com/OAcBmwW.png) if the first is **not** available.

cramatsu replied:
https://imgur.com/ZPIEalu
Is it ok?

cramatsu replied:
And another one that was made by me
https://imgur.com/5DZ70gr

plskz replied:
use [imgur.com/ZEnNmek](https://i.imgur.com/ZEnNmek.png)

plskz commented on websites/F/Foxford/dist/metadata.json:
```diff
@@ -0,0 +1,19 @@
+	"thumbnail": "https://i.imgur.com/41WDw6H.jpeg",
+	"color": "#ff9300",
+	"category": "other",
+	"tags": ["school", "online-shool", "courses"]
```
```suggestion
	"tags": ["school", "learning", "courses"]
```

cramatsu replied:
Thank you for noticing my typo

DarkVillager reviewed (approved):
Seems fine to me

RisingSunLight42 reviewed (commented):
Maybe for every switch were you only edit the presence details you can create an object to store a pair of key/value correspond to the pathHas() and the details given to shorten the code.

cramatsu commented:
> Maybe for every switch were you only edit the presence details you can create an object to store a pair of key/value correspond to the pathHas() and the details given to shorten the code.

Not a bad idea, now I'll rewrite the code for this

cramatsu commented:
What do you think about this code?
Yes, I know that the DRY principle is violated in a couple of lines, 'cause it's raw

```ts
const presence = new Presence({
	clientId: "1019291803322679456",
});

function pathHas(pathName: string): boolean {
	return document.location.pathname.toLowerCase().includes(pathName);
}

const routes = {
	"conspects": "Просматривает конспект",
	"account": "Настраивает профиль",
	"externship": "Просматривает программу обучения",
	"calendar": "Смотрит календарь занятий",
	"daily-plan": "Смотрит план занятий",
	"objectives": "Просматривает задания",
	"progress": "Просматривает успеваемость",
	"dashboard": "Просматривает курсы",
} as const;

presence.on("UpdateData", async () => {
	const presenceData: PresenceData = {
		largeImageKey: "f1024",
	};

	switch (true) {
		case location.pathname.split("/").pop() in routes:
			presenceData.details =
				routes[
					location.pathname.split("/").pop() as unknown as keyof typeof routes
				];
			break;
		case pathHas("groups"):
			presenceData.details = "Просматривает урок";
			break;
		case pathHas("courses"):
			presenceData.state = document.querySelector(
				".Header_title__G-XGe"
			).textContent;
			presenceData.details = "Просматривает курс";
			break;
		case pathHas("users"):
			presenceData.details = "Просматривает профиль";
			presenceData.buttons = [
				{
					label: "Перейти",
					url: document.location.href,
				},
			];
			break;
		case pathHas("tasks"):
			presenceData.details = "Решает задачи";
			presenceData.buttons = [
				{
					label: "Перейти",
					url: document.location.href,
				},
			];
			break;
		default:
			presenceData.details = "Просматривает разделы сайта";
			break;
	}

	if (presenceData.details) presence.setActivity(presenceData);
	else presence.setActivity();
});
```

**About `group` route**
This route has the following structure: <protocol>://foxford.ru/groups/[0-9]

RisingSunLight42 commented:
Seems good for me

cramatsu commented:
Is there any other way to improve the code or leave it as it is.
_I just ran out of ideas._
